app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java

```diff
@@ -467,6 +622,55 @@ public void onCopyWikicodeClicked(){
         Toast.makeText(getContext(), getString(R.string.wikicode_copied), Toast.LENGTH_SHORT).show();
     }
 
+    @OnClick(R.id.dummy_category_edit_container)
+    public void onOutsideOfCategoryEditClicked() {
+        if (dummyCategoryEditContainer.getVisibility() == VISIBLE) {
+            dummyCategoryEditContainer.setVisibility(GONE);
+        }
+    }
+
+    @OnClick(R.id.categoryEditButton)
+    public void onCategoryEditButtonClicked(){
+        displayHideCategorySearch();
+    }
+
+    public void displayHideCategorySearch() {
+        if (dummyCategoryEditContainer.getVisibility() != VISIBLE) {
+            dummyCategoryEditContainer.setVisibility(VISIBLE);
+        } else {
+            dummyCategoryEditContainer.setVisibility(GONE);
+        }
+    }
+
+    @OnClick(R.id.update_categories_button)
+    public void onUpdateCategoriesClicked() {
+       updateCategories(categoryEditSearchRecyclerViewAdapter.getNewCategories());
+        displayHideCategorySearch();
+    }
+
+    @OnClick(R.id.cancel_categories_button)
+    public void onCancelCategoriesClicked() {
+        displayHideCategorySearch();
+    }
+
+    public void updateCategories(List<String> selectedCategories) {
+        Single<Boolean> resultSingle =reasonBuilder.getReason(media, null)
+            .flatMap((reason) -> categoryEditHelper.makeCategoryEdit(getContext(), media, selectedCategories, this));
```

Comment:
Here @ashishkumar468 , there is no need to build a reason. I just tried to mimic the same implementation as deleteHelper which does not raise exception. I don't know what I can put to this line, to be able to call flatMap. 

For other implementation, see line 747 of same file
```
    private void onDeleteClicked(Spinner spinner) {
```

Reply:
I could not understand @neslihanturan . What is the issue with this implementation?

Reply:
well, I don't need to buıild a reason as "reasonBuilder.getReason(media, null)". It is only a need for deletion request. I used that line just to be able to implement the same structure with line 747. But my rxjava knowledge doesnt help me to find what to put before .flatMap, if not needed. Does it clear now?